src/ocean/net/http/HttpServer.d

```diff
@@ -0,0 +1,232 @@
+/// else than `Not Found`
+private class HttpRequestHandler : TaskHttpConnectionHandler
+{
+    /// The http handler list
+    private HttpHandler[] handlers;
+
+    /// Log instance
+    private Logger log;
+
+    /**************************************************************************
+
+        Constructor
+
+        Params:
+            finalizer = finalizer callback of the select listener
+
+    **************************************************************************/
+
+    public this ( scope FinalizeDg finalizer, HttpHandler[] handlers )
+    {
+        super(finalizer, new HttpRequest(true), new HttpResponse,
+            [ HttpMethod.Get, HttpMethod.Put, HttpMethod.Post,
+              HttpMethod.Delete, HttpMethod.Options ]);
+        this.handlers = handlers;
+
+        this.log = Log.lookup("ocean.net.http.HttpServer.HttpRequestHandler");
+    }
+
+    /**************************************************************************
+
+        Method that handles the HTTP requests, it will call each
+
+        Params:
+            response = the response message that will be sent to the client
+
+        Returns:
+            HttpResponseCode.OK if the request was successfully handled, or
+            HttpResponseCode.BadRequest in case of an error
+
+    **************************************************************************/
+
+    override public HttpResponseCode handleRequest ( out const(char)[] response )
+    {
+        log.trace("Got HTTP request");
+
+        try
+        {
+            foreach ( handler; this.handlers )
+            {
+                auto status = handler(this.request, response);
+
+                if ( status != HttpResponseCode.NotFound )
+                {
+                    return status;
+                }
+            }
+        }
+        catch ( Exception e )
+        {
+            log.error("Can't handle the request: `{}` `{}` `{}`", this.request.uri_string, this.request.msg_body, e.message);
+            response = e.message;
+
+            return HttpResponseCode.BadRequest;
+        }
+
+        return HttpResponseCode.NotFound;
+    }
+
+    /**************************************************************************
+
+        Tells the request message body length.
+        This method should be overridden when a request message body is
+        expected. It is invoked when the message header is completely parsed.
+        The default behavior is expecting no request message body.
+
+        Returns:
+            the request message body length in bytes (0 indicates that no
+            request message body is expected)
+
+     **************************************************************************/
+
+    override protected size_t request_msg_body_length ( )
+    {
+        size_t size;
+
+        try
+        {
+            this.request.getUnsigned("Content-Length", size);
+        }
+        catch(Exception e)
+        {
+            log.error("{}", e.toString);
```

Comment:
`toString` allocates and walk the stack, hence it's a potential DoS vector. Usually we use `message` for this.